File: tcshow.py
#!/usr/bin/env python3
"""
Read-only analysis helpers for grandMA2 timecode XML — standard library only,
no Qt. Used by the GUI for the info panel and the tempo estimate, and handy on
its own.
"""
import re

FRAME_FMT = re.compile(r'frame_format="[^"]*?(\d+)')
TC_NAME   = re.compile(r'<Timecode\b[^>]*\bname="([^"]*)"')
SUBTRACK  = re.compile(r'<SubTrack\b')
EVENT     = re.compile(r'<Event\b[^>]*\btime="(\d+)"')
OBJ_NAME  = re.compile(r'<Object\b[^>]*\bname="([^"]*)"')
CUE_NAME  = re.compile(r'<Cue\b[^>]*\bname="([^"]*)"')


def read_show(path):
    """Return (has_bom, text). Text is decoded without the BOM."""
    raw = open(path, 'rb').read()
    bom = raw.startswith(b'\xef\xbb\xbf')
    return bom, (raw[3:] if bom else raw).decode('utf-8')


def get_fps(text):
    m = FRAME_FMT.search(text)
    if not m:
        raise ValueError('frame_format not found — is this a grandMA2 timecode XML?')
    return int(m.group(1))


def frames_to_tc(fr, fps):
    s, f = divmod(int(fr), fps)
    h, s = divmod(s, 3600)
    m, s = divmod(s, 60)
    return f"{h:02}:{m:02}:{s:02}:{f:02}"


def events_by_subtrack(text):
    """List per subtrack: [(subtrack_index, [frame, ...]), ...] in document order."""
    eol = '\r\n' if '\r\n' in text else '\n'
    result, cur = [], None
    for line in text.split(eol):
        if SUBTRACK.search(line):
            cur = []
            result.append(cur)
        elif cur is not None:
            m = EVENT.search(line)
            if m:
                cur.append(int(m.group(1)))
    return result


def summary(text):
    """Quick facts for the info panel."""
    fps = get_fps(text)
    subs = events_by_subtrack(text)
    times = [t for s in subs for t in s]
    nm = TC_NAME.search(text)
    first, last = (min(times), max(times)) if times else (0, 0)
    return {
        'fps': fps,
        'name': nm.group(1) if nm else '(unnamed)',
        'n_events': len(times),
        'n_subtracks': len(subs),
        'first_frame': first,
        'last_frame': last,
        'first_tc': frames_to_tc(first, fps),
        'last_tc': frames_to_tc(last, fps),
    }


def estimate_beat(text):
    """
    Rough tempo estimate from the cue grid: look at moments where >=3 subtracks
    fire together (almost always downbeats), and find the frame length that best
    divides the gaps between them. Returns (beat_frames, bpm) or None.

    Heuristic — meant as a hint. Always let the user override with the real BPM.
    """
    fps = get_fps(text)
    subs = events_by_subtrack(text)
    hits = {}
    for si, times in enumerate(subs):
        for t in times:
            hits.setdefault(t, set()).add(si)
    struct = sorted(t for t, s in hits.items() if len(s) >= 3)
    gaps = [b - a for a, b in zip(struct, struct[1:])]
    gaps = [g for g in gaps if 10 <= g <= 700]      # drop sub-beat chases and long sections
    if len(gaps) < 3:
        return None
    best_b, best_err = None, 1e9
    b = 10.0
    while b <= 40.0:                                  # ~45..180 BPM at 30fps
        err = 0.0
        for g in gaps:
            r = g % b
            err += min(r, b - r) / b
        err /= len(gaps)
        if err < best_err:
            best_err, best_b = err, b
        b += 0.1
    if best_b is None or best_err > 0.18:             # too noisy -> no confident grid
        return None
    # Octave correction: the minimizer is biased toward small divisors (an eighth
    # fits as well as the quarter). Fold the implied tempo into a musical range so
    # we report the felt beat, not a harmonic of it.
    bpm = fps * 60.0 / best_b
    while bpm >= 140.0:
        bpm /= 2.0
    while bpm < 70.0:
        bpm *= 2.0
    return fps * 60.0 / bpm, bpm


def lanes(text):
    """
    Per-subtrack lanes for the timeline:
    [{'name': <object/track label>, 'events': [(frame, cue_name), ...]}, ...]
    """
    eol = '\r\n' if '\r\n' in text else '\n'
    out, cur, pending, obj = [], None, None, None
    for line in text.split(eol):
        mo = OBJ_NAME.search(line)
        if mo:
            obj = mo.group(1)
            continue
        if SUBTRACK.search(line):
            cur = {'name': obj or f'Track {len(out) + 1}', 'events': []}
            out.append(cur)
            obj, pending = None, None
            continue
        if cur is None:
            continue
        me = EVENT.search(line)
        if me:
            pending = int(me.group(1))
            continue
        mc = CUE_NAME.search(line)
        if mc and pending is not None:
            cur['events'].append((pending, mc.group(1)))
            pending = None
    return out

File: web/src/core/tcshow.ts
// Read-only analysis of grandMA2 timecode XML — port of tcshow.py.
// Drives the info line, the timeline lanes, and the BPM estimate.
import { framesToTc } from "./frames.ts";

const FRAME_FMT = /frame_format="[^"]*?(\d+)/;
const TC_NAME = /<Timecode\b[^>]*\bname="([^"]*)"/;
const SUBTRACK = /<SubTrack\b/;
const EVENT = /<Event\b[^>]*\btime="(\d+)"/;
const OBJ_NAME = /<Object\b[^>]*\bname="([^"]*)"/;
const CUE_NAME = /<Cue\b[^>]*\bname="([^"]*)"/;
const BOM = "﻿";

export interface ShowSummary {
  fps: number;
  name: string;
  nEvents: number;
  nSubtracks: number;
  firstFrame: number;
  lastFrame: number;
  firstTc: string;
  lastTc: string;
}

export interface Lane {
  name: string;
  events: Array<[number, string]>; // (frame, cue name)
}

/** Decode an uploaded file: strip a leading UTF-8 BOM, remember whether it was there. */
export function decodeShow(buf: ArrayBuffer): { hasBom: boolean; text: string } {
  const bytes = new Uint8Array(buf);
  const hasBom = bytes.length >= 3 && bytes[0] === 0xef && bytes[1] === 0xbb && bytes[2] === 0xbf;
  const body = hasBom ? bytes.subarray(3) : bytes;
  const text = new TextDecoder("utf-8").decode(body);
  return { hasBom, text };
}

/** Re-encode for download: prepend the BOM iff the original had one. */
export function encodeShow(text: string, hasBom: boolean): Uint8Array {
  const body = new TextEncoder().encode(text);
  if (!hasBom) return body;
  const out = new Uint8Array(body.length + 3);
  out.set([0xef, 0xbb, 0xbf], 0);
  out.set(body, 3);
  return out;
}

export function getFps(text: string): number {
  const m = FRAME_FMT.exec(text);
  if (!m) throw new Error("frame_format not found — is this a grandMA2 timecode XML?");
  return parseInt(m[1], 10);
}

function eol(text: string): string {
  return text.includes("\r\n") ? "\r\n" : "\n";
}

/** Frames per subtrack, in document order. */
export function eventsBySubtrack(text: string): number[][] {
  const result: number[][] = [];
  let cur: number[] | null = null;
  for (const line of text.split(eol(text))) {
    if (SUBTRACK.test(line)) {
      cur = [];
      result.push(cur);
    } else if (cur !== null) {
      const m = EVENT.exec(line);
      if (m) cur.push(parseInt(m[1], 10));
    }
  }
  return result;
}

export function summary(text: string): ShowSummary {
  const fps = getFps(text);
  const subs = eventsBySubtrack(text);
  const times: number[] = [];
  for (const s of subs) for (const t of s) times.push(t);
  const nm = TC_NAME.exec(text);
  const first = times.length ? Math.min(...times) : 0;
  const last = times.length ? Math.max(...times) : 0;
  return {
    fps,
    name: nm ? nm[1] : "(unnamed)",
    nEvents: times.length,
    nSubtracks: subs.length,
    firstFrame: first,
    lastFrame: last,
    firstTc: framesToTc(first, fps),
    lastTc: framesToTc(last, fps),
  };
}

/**
 * Rough tempo from the cue grid: moments where >=3 subtracks fire together are
 * almost always downbeats; find the frame length that best divides the gaps.
 * Returns { beatFrames, bpm } or null. A hint — always overridable.
 */
export function estimateBeat(text: string): { beatFrames: number; bpm: number } | null {
  const fps = getFps(text);
  const subs = eventsBySubtrack(text);
  const hits = new Map<number, Set<number>>();
  subs.forEach((times, si) => {
    for (const t of times) {
      let s = hits.get(t);
      if (!s) hits.set(t, (s = new Set()));
      s.add(si);
    }
  });
  const struct = [...hits.entries()].filter(([, s]) => s.size >= 3).map(([t]) => t).sort((a, b) => a - b);
  let gaps: number[] = [];
  for (let i = 1; i < struct.length; i++) gaps.push(struct[i] - struct[i - 1]);
  gaps = gaps.filter((g) => g >= 10 && g <= 700);
  if (gaps.length < 3) return null;

  let bestB: number | null = null;
  let bestErr = 1e9;
  for (let b = 10.0; b <= 40.0; b += 0.1) {
    let err = 0;
    for (const g of gaps) {
      const r = g % b;
      err += Math.min(r, b - r) / b;
    }
    err /= gaps.length;
    if (err < bestErr) {
      bestErr = err;
      bestB = b;
    }
  }
  if (bestB === null || bestErr > 0.18) return null;

  // Octave-correct into a musical range so we report the felt beat.
  let bpm = (fps * 60.0) / bestB;
  while (bpm >= 140.0) bpm /= 2.0;
  while (bpm < 70.0) bpm *= 2.0;
  return { beatFrames: (fps * 60.0) / bpm, bpm };
}

/** Per-subtrack lanes for the timeline. */
export function lanes(text: string): Lane[] {
  const out: Lane[] = [];
  let cur: Lane | null = null;
  let pending: number | null = null;
  let obj: string | null = null;
  for (const line of text.split(eol(text))) {
    const mo = OBJ_NAME.exec(line);
    if (mo) {
      obj = mo[1];
      continue;
    }
    if (SUBTRACK.test(line)) {
      cur = { name: obj || `Track ${out.length + 1}`, events: [] };
      out.push(cur);
      obj = null;
      pending = null;
      continue;
    }
    if (cur === null) continue;
    const me = EVENT.exec(line);
    if (me) {
      pending = parseInt(me[1], 10);
      continue;
    }
    const mc = CUE_NAME.exec(line);
    if (mc && pending !== null) {
      cur.events.push([pending, mc[1]]);
      pending = null;
    }
  }
  return out;
}

export { BOM };
